xs/src/libslic3r/PerimeterGenerator.cpp

```diff
@@ -6,6 +6,23 @@
 
 namespace Slic3r {
 
+int
+PerimeterGenerator::num_loops(int perimeters, int extra_perimeters)
+{
+    int generated_perimeters = perimeters + extra_perimeters - 1;
+    long min_shell_thickness = scale_(this->config->min_shell_thickness);
+    auto shell_thickness = this->ext_perimeter_flow.scaled_width();
+
+    shell_thickness += ((perimeters+extra_perimeters)-1) * this->perimeter_flow.scaled_width();
+    
+    if (shell_thickness < min_shell_thickness) {
+        shell_thickness = min_shell_thickness - this->ext_perimeter_flow.scaled_width();
+        generated_perimeters = ceil((float)shell_thickness/this->ext_perimeter_flow.scaled_width());
+    }
```

Comment:
Why calling `this->ext_perimeter_flow.scaled_width()` and `/this->ext_perimeter_flow.scaled_width()` twice?

By the way, I would simplify the logic:
1. Calculate how many loops are needed for the requested min_shell_thickness (considering one ext_perimeter followed by many normal perimeters, of course).
2. Increase loop number until min_shell_thickness is covered.

Also, write some comments.